src/main/scala/common/master_adapter.scala

```diff
@@ -0,0 +1,128 @@
+  val a_signed_reg = RegInit(false.B)
+  // Request register - store the request when the input port fires to avoid value changes when sending the TileLink request
+  val req_address_reg = RegInit(0.U(io.dport.req.bits.addr.getWidth.W))
+  val req_size_reg = RegInit(0.U(2.W))
+  val req_data_reg = RegInit(0.U(io.dport.req.bits.data.getWidth.W))
+
+  // Sign logic
+  // To convert MemPortIO type to sign and size in TileLink format: subtract 1 from type, then take inversed MSB as signedness
+  // and the remaining two bits as TileLink size
+  val a_signed = ~(io.dport.req.bits.typ - 1.U)(2)
+  val a_size = (io.dport.req.bits.typ - 1.U)(1, 0)
+
+  // State logic
+  when (state === s_ready && io.dport.req.valid) {
+    state := s_active
+    req_address_reg := io.dport.req.bits.addr
+    req_size_reg := a_size
+    req_data_reg := io.dport.req.bits.data
+  }
+  when (state === s_active && tl_out.a.fire()) {
+    state := s_inflight
+  }
+  when (state === s_inflight && tl_out.d.fire()) {
+    state := s_ready
+  }
+  tl_out.a.valid := state === s_active
+  tl_out.d.ready := true.B
+  io.dport.req.ready := state === s_ready
+  io.dport.resp.valid := tl_out.d.valid
+
+  // Bookkeeping
+  when (tl_out.a.fire()) {
+    a_address_reg := io.dport.req.bits.addr
+    a_signed_reg := a_size
+  }
+
+  // Build "Get" message
+  val (legal_get, get_bundle) = edge.Get(0.U, req_address_reg, req_size_reg)
+  // Build "Put" message
+  val (legal_put, put_bundle) = edge.Put(0.U, req_address_reg, req_size_reg, req_data_reg)
+
+  // DEBUG
+  when (tl_out.a.fire()) {
+    printf("DEBUG MESSAGE: A-chan size is %x, with address %x\n", tl_out.a.bits.size, tl_out.a.bits.address)
+    printf("DEBUG MESSAGE: Original typ: %b\n", io.dport.req.bits.typ)
+  }
+
+  // Connect Channel A bundle
+  tl_out.a.bits := Mux(io.dport.req.bits.fcn === M_XRD, get_bundle, put_bundle)
+
+  // Connect Channel D bundle (read result)
+  io.dport.resp.bits.data := new LoadGen(tl_out.d.bits.size, a_signed_reg, a_address_reg, tl_out.d.bits.data, false.B, conf.xprlen / 8).data
+
+  // Handle error
+  val legal_op = Mux(io.dport.req.bits.fcn === M_XRD, legal_get, legal_put)
+  val resp_xp = tl_out.d.bits.corrupt | tl_out.d.bits.denied
+  // Since the core doesn't have an external exception port, we have to kill it
+  assert(legal_op | !tl_out.a.valid, "Illegal operation")
+  assert(!resp_xp | !tl_out.d.valid, "Responds exception")
+
+  // Tie off unused channels
+  tl_out.b.valid := false.B
+  tl_out.c.ready := true.B
+  tl_out.e.ready := true.B
+}
+
+// This class allows the next memory request to be sent when the response of the previous request come back. 
+class SameCycleRequestBuffer(implicit val conf: SodorConfiguration) extends Module {
+  val io = IO(new Bundle() {
+    val in = Flipped(new MemPortIo(data_width = conf.xprlen))
+    val out = new MemPortIo(data_width = conf.xprlen)
+  })
+  
+  io.out.req <> io.in.req
+  io.in.resp := RegNext(io.out.resp)
```

Comment:
The register driving `io.in.resp.valid` should be reset to `false.B`.
```suggestion
  io.in.resp := Pipe(io.out.resp)
```

Reply:
Done